__tests__/utils/flattenProperties.test.js

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
+ * the License. A copy of the License is located at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+var flattenProperties = require('../../lib/utils/flattenProperties');
+
+describe('utils', () => {
+  describe('flattenProperties', () => {
+
+    it('should return an empty array', () => {
+      var ret = flattenProperties({});
+      expect(ret).toEqual([]);
+    });
+
+    it('should return the same array', () => {
+      var to_ret = [];
+      var ret = flattenProperties({}, to_ret);
+      expect(ret).toBe(ret);
+    });
+
+    it('should return leaf node values as an array', () => {
+      var properties = {
+        'black': {
+          'value': '#000000'
+        },
+        'white': {
+          'value': '#FFFFFF'
+        }
+      };
+      var expected_ret = [
+        { 'value': '#000000' },
+        { 'value': '#FFFFFF' }
+      ];
+      var sortedExpectedValues = expected_ret.map(v => v.value).sort();
+      var ret = flattenProperties(properties);
+      var sortedReturnedValues = ret.map(v => v.value).sort();
+      expect(sortedReturnedValues).toEqual(sortedExpectedValues);
+    });
+
+    it('should return nested leaf node values as an array', () => {
+      var properties = {
+        'color': {
+          'black': {
+            'value': '#000000'
+          },
+          'white': {
+            'value': '#FFFFFF'
+          }
+        }
+      };
+      var expected_ret = [
+        { 'value': '#000000' },
+        { 'value': '#FFFFFF' }
+      ];
+      var sortedExpectedValues = expected_ret.map(v => v.value).sort();
+      var ret = flattenProperties(properties);
+      var sortedReturnedValues = ret.map(v => v.value).sort();
+      expect(sortedReturnedValues).toEqual(sortedExpectedValues);
+    });
```

Comment:
For this last test, instead of testing the values of the value key, I think we should test the objects against each other. This will enable us to add in details to the objects later and make sure they land correctly. Maybe something like:

```
it('should return nested leaf node values as an array', () => {
      var properties = {
        'color': {
          'black': {
            'value': '#000000'
          },
          'white': {
            'value': '#FFFFFF'
          }
        }
      };
      var expected_ret = [
        properties.color.black,
        properties.color.white
      ];
      var sortFn = function(a,b) {
            if (a.value > b.value)
                  return 1;
            else if (b.value > a.value)
                  return -1;
            else
                  return 0;
      };
      var sortedExpectedValues = expected_ret.sort(sortFn);
      var ret = flattenProperties(properties);
      var sortedReturnedValues = ret.sort(sortFn);
      expect(JSON.stringify(sortedReturnedValues)).toEqual(JSON.stringify(sortedReturnedValues));
    });
  });
```

Reply:
Yes it makes sense! Shall I add this and update my pull request?